src/spaces/openstreetmap.jl

```diff
@@ -0,0 +1,402 @@
+export OpenStreetMapSpace, OSMPos, OSMSpace
+export osm_random_direction, osm_plan_route, osm_map_coordinates, osm_road_length
+
+struct OpenStreetMapSpace <: DiscreteSpace
+    m::OpenStreetMapX.MapData
+    edges::Vector{CartesianIndex{2}} #So far, only helpful for osm_random_direction
+    s::Vector{Vector{Int}}
+end
+
+"""
+    OpenStreetMapSpace(path::AbstractString; kwargs...)
+Create a space residing on the Open Street Map (OSM) file provided via `path`.
+
+The abbreviation `OSMSpace` may be used interchangably.
+
+Much of the functionality of this space is provided by interfacing with
+[OpenStreetMapX.jl](https://github.com/pszufe/OpenStreetMapX.jl), for example the two
+keyword arguments `use_cache = false` and `trim_to_connected_graph = true` are
+passed into the [`OpenStreetMapX.get_map_data`](@ref) function.
+For details on how to obtain an OSM file for your use case, consult the OpenStreetMapX.jl
+Readme.
+"""
+function OpenStreetMapSpace(
+    path::AbstractString;
+    use_cache = false,
+    trim_to_connected_graph = true,
+)
+    m = get_map_data(
+        path; use_cache, trim_to_connected_graph,
+    )
+    agent_positions = [Int[] for i in 1:nv(m.g)]
+    return OpenStreetMapSpace(m, findall(!iszero, m.w), agent_positions)
+end
+
+function Base.show(io::IO, s::OpenStreetMapSpace)
+    print(
+        io,
+        "OpenStreetMapSpace with $(length(s.m.roadways)) roadways and $(length(s.m.intersections)) intersections",
+    )
+end
+
+const OSMSpace = OpenStreetMapSpace
+
+struct OSMPos
+    p::Tuple{Int, Int, Float64}
+end
+Base.getindex(o::OSMPos, i::Int) = o.p[i]
+
+"""
+    OSMPos(start::Int, finish::Int = start, road::Float64 = 0)
+
+A helper to provide the correct `pos` format for [`AbstractAgent`](@ref) structs which
+are used in conjuction with [`OpenStreetMapSpace`](@ref).
+It represents the position of the agent on a road connecting the two nodes of the map
+`start, finish` having already covered `road` meters along the road.
+"""
+OSMPos(start::Int, finish::Int = src, p::Float64 = 0.0) = OSMPos((start, finish, p))
+
+#######################################################################################
+# Custom functions for OSMSpace
+#######################################################################################
+
+# EXPORTED
+"""
+    osm_random_direction(model::ABM{A,OpenStreetMapSpace})
+
+Similar to `random_position`, but rather than providing only intersections, this method
+returns a location somewhere on a road heading in a random direction.
+"""
```

Comment:
It is not completely clear what this function returns. We should add an arrow `-> OSMPos`. But the function name also doesn't help. It doesn't feel liike this function returns a valid position instance for the map space, but perhaps something else. Given that I've added `osm` in front, we can maybe make it `osm_random_position` instead?

Reply:
"Position" in OSMX.jl is a node on a graph - which for users would translate to intersections on a map. #332 uses `random_position` to choose one of these intersections at random, but therefore cannot provide all positions on the map. That function is still useful in many cases though, so I kept it as-is.

This new one, it takes a random *edge* of the graph (which translates to road), and chooses a random position on that road, as well as a heading (or direction of travel)). In that sense, this is not an `osm_random_position`, it's closer to `osm_random_vector`, but doesn't have a velocity associated with it, so "direction" or "heading" are better qualifiers.

Reply:
See the example, we have
```julia
start = osm_random_direction(model) # Somewhere on a road
finish = random_position(model) # At an intersection
```
as a contrast. We could rename `random_position` to `osm_random_intersection`, then reconsider what `osm_random_direction` should be called.